teps/0147-dynamic-matrix.md

```diff
@@ -0,0 +1,210 @@
+          - name: build-3
+            params:
+              - name: IMAGE
+                value: "image-3"
+              - name: DOCKERFILE
+                value: "path/to/Dockerfile3"
+    - ...
+```
+
+This `pipeline` is part of a catalog and shared across multiple teams. Now, the teams are building their own applications
+using this `pipeline` and different teams have different number of images to built as part of their application. This way
+of specifying `matrix` combinations in a `pipeline` is constant and can not be utilized for an application with
+variety of images.
+
+## Proposal
+
+We propose adding a field - `strategy` of type `string` - within the `matrix` section. This allows pipeline authors to
+maintain a single pipeline in a catalog and allow the users to specify the list of explicit combinations of `params`
+dynamically through `param` (`pipelineRun` or a trigger template) or a task result of a parent task.
+
+The `matrix.strategy` is of type `string`  with a JSON payload `{"include": a list of combinations}`. This aligns with
+`matrix.include` syntax and allows to specify dynamic list of combinations.
+
+The `Pipeline` addressing this use case will be defined as shown below:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: pipeline-to-build-images-from-a-single-repo
+spec:
+  params:
+    - name: matrix-strategy
+  workspaces:
+    - name: shared-workspace
+  tasks:
+    - ...
+    - name: kaniko-build
+      taskRef:
+        name: kaniko
+      workspaces:
+        - name: source
+          workspace: shared-workspace
+      matrix:
+        strategy: $(params.matrix-strategy)
+```
+
+A `PipelineRun` can be created to execute the above `pipeline` to build 1 image:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  generateName: pipelineRun-
+spec:
+  pipelineRef:
+    name: pipeline-to-build-images-from-a-single-repo
+  params:
+    - name: matrix-strategy
+      value: "{\"include\":[{\"IMAGE\":\"image-1\",\"DOCKERFILE\":\"path/to/Dockerfile1\"}]}"
+```
+
+`PipelineRun` can be created to execute the above `pipeline` to build 3 images:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  generateName: pipelineRun-
+spec:
+  pipelineRef:
+    name: pipeline-to-build-images-from-a-single-repo 
+  params:
+    - name: matrix-strategy
+      value: "{\"include\":[{\"IMAGE\":\"image-1\",\"DOCKERFILE\":\"path/to/Dockerfile1\"},{\"IMAGE\":\"image-2\",\"DOCKERFILE\":\"path/to/Dockerfile2\"},{\"IMAGE\":\"image-3\",\"DOCKERFILE\":\"path/to/Dockerfile3\"}]}"
+```
+
+This syntax is inspired by the GitHub actions syntax similar to the other sections of `matrix`.
+
+[GitHub actions](https://docs.github.com/en/actions/learn-github-actions/expressions#example-returning-a-json-object)
+does support specifying a list of jobs in which a JSON payload is set in one job and passed to the next job using
+`output` and `fromJSON`.
```

Comment:
As mentioned in the Tekton API workgroup it might be worth investigating if CEL expressions can help w/ creating something like `fromJSON` for the Tekton implementation well

Reply:
Thanks @aaron-prindle, I have added `CEL` support as an additional feature along with a `JSON` payload such that we are not limited to a single expression language.

https://github.com/tektoncd/community/pull/1081/files#diff-2f6a5f6cb9769c1aabe77ce0cd3e03608bf1e283be787d0d4eb53c957176bbf6R352

Reply:
One other alternative would be to build out our array/object parameter support better so that we could pass the strategy value as an array of objects instead of a string

Reply:
@dibyom would you consider that as a follow-up to this TEP?

Reply:
Done, thanks @dibyom!